### PR TITLE
Migrate ModelBrowser to Kotlin

### PR DESCRIPTION
**Pull request by ShridharGoel**

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 1]
```diff
@@ -14,608 +14,501 @@
...
-        @Override
-        public void actualOnPreExecute(@NonNull ModelBrowser browser) {
-            browser.showProgressBar();
+        override fun actualOnPreExecute(context: ModelBrowser?) {
+            context?.showProgressBar()
         }
 
-        @Override
-        public void actualOnPostExecute(@NonNull ModelBrowser browser, Pair<List<Model>, ArrayList<Integer>> result) {
+        override fun actualOnPostExecute(context: ModelBrowser?, result: Pair<List<Model?>?, ArrayList<Int?>?>?) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:97]: Cleanup: rename `context` in the base class to `activity`.

(and see if we can make it non-null)

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 2]
```diff
@@ -14,608 +14,501 @@
...
+        menuInflater.inflate(R.menu.model_browser, menu)
+        return true
     }
 
-    @Override
-    public boolean onOptionsItemSelected(MenuItem item) {
-        int itemId = item.getItemId();
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        val itemId = item.itemId
         if (itemId == android.R.id.home) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:171]: cleanup: replace with `when`

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 3]
```diff
@@ -14,608 +14,501 @@
...
             } else {
-                //New model
-                //Model that is being cloned
-                Model oldModel = mModels.get(position - nbStdModels).deepClone();
-                Model newModel = StdModels.BASIC_MODEL.add(mCol);
-                oldModel.put("id", newModel.getLong("id"));
-                model = oldModel;
+                // New model
+                // Model that is being cloned
+                val oldModel = mModels!![position - nbStdModels].deepClone()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:307]: scope function?

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 4]
```diff
@@ -14,608 +14,501 @@
...
             }
-            model.put("name", modelName);
-            mCol.getModels().update(model);
-            fullRefresh();
+            model.put("name", modelName)
+            mCol!!.models.update(model)
+            fullRefresh()
         } else {
-            showToast(getResources().getString(R.string.toast_empty_name));
+            showToast(resources.getString(R.string.toast_empty_name))
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:316]: cleanup: invert + return early

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 5]
```diff
@@ -14,608 +14,501 @@
...
-                ModelBrowser.this.showDialogFragment(d);
-            } catch (ConfirmModSchemaException e) {
-                e.log();
-                ConfirmationDialog c = new ConfirmationDialog();
-                c.setArgs(getResources().getString(R.string.full_sync_confirmation));
-                c.setConfirm(confirm);
-                c.setCancel(cancel);
-                showDialogFragment(c);
+                mCol!!.modSchema()
+                val d = ConfirmationDialog()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:361]: cleanup: rename

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 6]
```diff
@@ -14,608 +14,501 @@
...
+                d.setArgs(resources.getString(R.string.model_delete_warning))
+                d.setConfirm(confirm)
+                d.setCancel(cancel)
+                this@ModelBrowser.showDialogFragment(d)
+            } catch (e: ConfirmModSchemaException) {
+                e.log()
+                val c = ConfirmationDialog()
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:368]: ditto

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 7]
```diff
@@ -14,608 +14,501 @@
...
+import com.ichi2.libanki.StdModels
+import com.ichi2.libanki.Utils
+import com.ichi2.ui.FixedEditText
+import com.ichi2.utils.KotlinCleanup
+import com.ichi2.widget.WidgetStatus.update
+import timber.log.Timber
+import java.lang.RuntimeException
+import java.util.ArrayList
+
+@KotlinCleanup("Try converting variables to be non-null wherever possible")
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:53]: + standard in-IDE cleanup

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 8]
```diff
@@ -14,608 +14,501 @@
...
-        }
-
-        public int getCount() {
-            return mCount;
-        }
-
-        @Override
-        public @NonNull String toString() {
-            return getName();
+    class DisplayPair(val name: String, val count: Int) {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:484]: Beautiful

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 9]
```diff
@@ -14,608 +14,501 @@
...
-        @NonNull
-        @Override
-        public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
-            DisplayPair item = getItem(position);
-
-            if (convertView == null) {
-                convertView = LayoutInflater.from(getContext()).inflate(R.layout.model_browser_list_item, parent, false);
+    inner class DisplayPairAdapter(context: Context?, items: ArrayList<DisplayPair>?) : ArrayAdapter<DisplayPair?>(context!!, R.layout.model_browser_list_item, R.id.model_list_item_1, items!!.toList()) {
+        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
+            var _convertView = convertView
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:495]: `?:`

**Review by david-allison — COMMENTED** [review]
Looks great! Few more cleanups

**Review by Arthur-Milchior — COMMENTED** [review]
there is a lot of nonNull context that become nullable after the translation. If it's not avoidable, could you at least explain why in the commit message?
Otherwise, I'd really prefer to avoid this change.

I'll read entirely once this is answered, as I fear that it may require a lot of changes anyway

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt [hunk 10]
```diff
@@ -14,608 +14,501 @@
...
-    private static class LoadingModelsHandler extends TaskListenerWithContext<ModelBrowser, Void, Pair<List<Model>, ArrayList<Integer>>> {
-        public LoadingModelsHandler(ModelBrowser browser) {
-            super(browser);
-        }
 
-        @Override
-        public void actualOnCancelled(@NonNull ModelBrowser browser) {
-            browser.hideProgressBar();
+    private class LoadingModelsHandler(browser: ModelBrowser?) : TaskListenerWithContext<ModelBrowser?, Void?, Pair<List<Model?>?, ArrayList<Int?>?>?>(browser) {
+        override fun actualOnCancelled(context: ModelBrowser?) {
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:89]: "CTX" in the base class needs a `: Any` constraint -

[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:89]: @david-allison I don't know whether you are answering my question or adding an extra note for @ShridharGoel

[the remark this answers is not in the document]
Comment (ShridharGoel) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:89]: I'll update this in another PR.

File: AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt (hunk 9 above)
Comment (ShridharGoel) [reply AnkiDroid/src/main/java/com/ichi2/anki/ModelBrowser.kt:495]: Can you clarify this?

**Review by david-allison — APPROVED** [review]
Comments addressed! Looks good to me

**Review by mikehardy — APPROVED** [review]
LGTM w/KotlinCleanups as proposed

**Pull request merged**